### PR TITLE
feat: metrics query range v3

### LINKED ISSUE #2241: (metrics): query range v3

### PR DESCRIPTION
**Pull request by srikanthccv**
Fixes #2241

### CODE REVIEW

File: pkg/query-service/app/http_handler.go [hunk 1]
```diff
@@ -101,6 +104,17 @@ func NewAPIHandler(opts APIHandlerOpts) (*APIHandler, error) {
...
+	builderOpts := queryBuilderOptions{
+		BuildMetricQuery: metricsv3.PrepareMetricQuery,
+		BuildTraceQuery: func(start, end, step int64, queryType v3.QueryType, panelType v3.PanelType, bq *v3.BuilderQuery) (string, error) {
+			return "", errors.New("not implemented")
+		},
+		BuildLogQuery: func(start, end, step int64, queryType v3.QueryType, panelType v3.PanelType, bq *v3.BuilderQuery) (string, error) {
+			return "", errors.New("not implemented")
+		},
+	}
+	aH.queryBuilder = NewQueryBuilder(builderOpts)
```
Comment (srikanthccv) [inline pkg/query-service/app/http_handler.go:116]: We haven't discussed this earlier as we did for other parts, making this ready for review to get started. Each query needs to be translated to ClickHouse SQL, and a formula will be based on the SQL of the individual queries. The `Build{Trace/Log/Metric}Query` handles the individual query. Once the simple queries are prepared, we use the govaluate to parse the formula and substitute the variable with its corresponding query. For instance

A sample request could be

```json
{
  "start": 1678650969000,
  "end": 1678652769000,
  "step": 60,
  "variables": {
    "SIGNOZ_START_TIME": 1678650969000,
    "SIGNOZ_END_TIME": 1678652769000
  },
  "dataSource": "metrics",
  "compositeQuery": {
    "queryType": "builder",
    "panelType": "graph",
    "builderQueries": {
      "A": {
        "queryName": "A",
        "dataSource": "metrics",
        "aggregateOperator": "sum_rate",
        "aggregateAttribute": "signoz_calls_total",
        "tagFilters": {
          "items": [
            {
              "key": "service_name",
              "op": "IN",
              "value": [
                "frontend"
              ]
            }
          ],
          "op": "AND"
        },
        "groupBy": [
           "service_name",
          "operation"
        ],
        "expression": "A",
        "disabled": false
      },
      "B": {
        "queryName": "B",
        "dataSource": "traces",
        "aggregateOperator": "rate",
        "aggregateAttribute": "sdsd",
        "tagFilters": {
          "op": "AND",
          "items": [
            {
              "id": "b4213c45",
              "key": "service_name",
              "op": "IN",
              "value": [
                "frontend"
              ]
            }
          ]
        },
        "groupBy": [
          "operation"
        ],
        "expression": "B",
        "disabled": false
      },
      "F1": {
        "queryName": "F1",
        "dataSource": "metrics",
        "expression": "A/B",
        "disabled": false,
        "legend": ""
      }
    }
  }
}
```

So the result for queries `A` and `B` should match mostly, and the Formula should have `1` as a result in most points.

Comment (nityanandagohain) [reply pkg/query-service/app/http_handler.go:116]: By this you are saying, we are generating the SQL for A & B and then creating a final SQL expression and then executing it right ?

Comment (srikanthccv) [reply pkg/query-service/app/http_handler.go:116]: Yes, that's correct, happy to hear any other ideas. Also relevant point https://github.com/SigNoz/signoz/issues/2376#issuecomment-1451078909.

Comment (makeavish) [reply pkg/query-service/app/http_handler.go:116]: So we aren't querying separately for A & B and then applying Formula over them in query service, which is currently being done in the v2 query range API?

Comment (srikanthccv) [reply pkg/query-service/app/http_handler.go:116]: That's not how it is currently being done in query service. The current query range uses ClickHouse to do the work.

Comment (makeavish) [reply pkg/query-service/app/http_handler.go:116]: > the Formula should have 1 as a result in most points.
I don't understand this statement, please explain further.

Comment (srikanthccv) [reply pkg/query-service/app/http_handler.go:116]: That was just an example. The first query `A` calculates the rate from the span metrics and the second query `B` calculates the rate from the traces so the formula `F1` = `A/B` should result in values being 1.

**Comment by srikanthccv** [comment]
I am happy to split this further if you think it will help to you to review

**Comment by makeavish** [comment]
> I am happy to split this further if you think it will help to you to review

I am waiting to integrate https://github.com/SigNoz/signoz/pull/2453 changes in this PR

**Review by nityanandagohain — APPROVED** [review]
Haven't gone through  `metrics/v3/query_builder.go`, but the common implementation part looks good to me.

**Review by makeavish — APPROVED** [review]

**Pull request merged**